src/query/functions/src/scalars/map.rs

```diff
@@ -156,4 +158,112 @@ pub fn register(registry: &mut FunctionRegistry) {
             |map, _| map.values
         ),
     );
+
+    registry.register_2_arg_core::<NullableType<EmptyMapType>, NullableType<EmptyMapType>, EmptyMapType, _, _>(
+        "map_cat",
+        |_, _, _| FunctionDomain::Full,
+        |_, _, _| Value::Scalar(()),
+    );
+
+    registry.register_passthrough_nullable_2_arg(
+        "map_cat",
+        |_, domain1, domain2| {
+            FunctionDomain::Domain(match (domain1, domain2) {
+                (Some((key_domain1, val_domain1)), Some((key_domain2, val_domain2))) => Some((
+                    key_domain1.merge(key_domain2),
+                    val_domain1.merge(val_domain2),
+                )),
+                (Some(domain1), None) => Some(domain1).cloned(),
+                (None, Some(domain2)) => Some(domain2).cloned(),
+                (None, None) => None,
+            })
+        },
+        vectorize_with_builder_2_arg::<
+            MapType<GenericType<0>, GenericType<1>>,
+            MapType<GenericType<0>, GenericType<1>>,
+            MapType<GenericType<0>, GenericType<1>>,
+        >(|lhs, rhs, output_map, _| {
+            let capacity = lhs.len();
+            let mut lhs_key_set: StackHashSet<u128, 16> = StackHashSet::with_capacity(capacity);
+
```

Comment:
we don't need `StackHashSet` here, we can use a `HashMap` to keep the duplicate keys, for example:
```rust
            let mut dup_map = HashMap::new();
            for (k1, _) in input_map1.iter() {
                for (k2, v2) in input_map2.iter() {
                    if k1 == k2 {
                        dup_map.insert(k2, v2);
                        break;
                    }
                }
            }
```
then we can add `input_map1` and `input_map2` to the builder.
```rust
            let mut builder = ArrayType::create_builder(capacity, ctx.generics);
            for (k1, v1) in input_map1.iter() {
                // check if `k1` is in the `dup_map`, if not, put `(k1, v1)` to the builder, otherwise, put `(k1, v2)` to the builder, we can get `v2` from the `dup_map`, which is the value in the `input_map2`.
                ...
            }
            for (k2, v2) in input_map2.iter() {
                // check if `k2` is in the `dup_map`, if not, put `(k2, v2)` to the builder.
                ...
            }
```
finally, add offsets to the builder and append it to the `output_map` 
```rust
            builder.commit_row();
            output_map.append_column(&builder.build());
```

Reply:
Good catch! Using `StackHashSet` wasn't the best call here. My bad.